002-es6-modules.md

```diff
@@ -0,0 +1,626 @@
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WHATWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+All exported values are declarative. That means that they are known after the file is parsed, but before it is evaluated.
+
+When creating a `DynamicModuleRecord`. The list of exports is frozen upon construction. No new exports may be added. No exports may be removed.
+
+## Algorithm
+
+When loading a file.
+
+1. Determine if file is ES or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing loading dependencies recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+  
+This still guarantees:
+
+* that ES module dependencies are all executed prior to the module itself
+* CJS modules have a full shape prior to being handed to ES modules
+* allows CJS modules to imperatively start the loading of other modules, including ES modules
+
+## Semantics
+
+### Determining if source is an ES Module
+
+A new filetype will be recognised, `.mjs` as ES modules. They will be treated as a different loading semantic but compatible with existing systems, just like `.node`, `.json`, or usage of `require.extension` (even though deprecated) are compatible. It would be ideal if we could register the filetype with IANA as an offical file type, see [TC39 issue](https://github.com/tc39/ecma262/issues/322). Though it seems this would need to go through the [IESG](https://www.ietf.org/iesg/) and it seems browsers are non-plussed on introducing a new MIME.
+
+The `.mjs` file extension will have a higher loading priority than `.js`.
```

Comment:
Can you clarify that "higher priority" means that a `require('./my_module')` will first search for `.mjs`?

Also, mind adding the information from https://github.com/nodejs/node-eps/issues/13#issuecomment-210596387 about why this extension was chosen (over `.es` and `.jsm`)? Think it's relevant enough to have directly in the EP.